### PR TITLE
SpatialPooler performance enhancement

### PR DESCRIPTION
**Pull request by ctrl-z-9000-times**
Rewrote method SpatialPooler::inhibitColumnsGlobal_. Previously it used an
iterative method with time complexity O(N^2).  New method uses algorithms from
the standard library with time complexity O(N log(N)).

Written by David McDougall

### CODE REVIEW

**Comment by dkeeney** [comment]
Cool, that seemed to have worked.

**Comment by ctrl-z-9000-times** [comment]
Thank you!

File: src/nupic/algorithms/SpatialPooler.cpp [hunk 1]
```diff
@@ -954,53 +955,33 @@ void SpatialPooler::inhibitColumns_(const vector<Real> &overlaps,
...
-    }
-  }
 
-  const UInt numActual = min(numDesired, (UInt)winners.size());
-  for (UInt i = 0; i < numActual; i++) {
-    activeColumns.push_back(winners[i].first);
+  activeColumns.reserve(numColumns_);
+  for(UInt i = 0; i < numColumns_; i++)
+    activeColumns.push_back(i);
+  auto compare = [&overlaps](const UInt &a, const UInt &b) -> bool
```
Comment (breznak) [inline src/nupic/algorithms/SpatialPooler.cpp]: please add a short doc on this improved implementation

File: src/nupic/algorithms/SpatialPooler.cpp [hunk 2]
```diff
@@ -954,53 +955,33 @@ void SpatialPooler::inhibitColumns_(const vector<Real> &overlaps,
...
+  std::nth_element(
+    activeColumns.begin(),
+    activeColumns.begin() + numDesired,
+    activeColumns.end(),
+    compare);
+  activeColumns.resize(numDesired);
+  std::sort(activeColumns.begin(), activeColumns.end(), compare);
+
+  // Remove sub-threshold winners
+  while(!activeColumns.empty()) {
```
Comment (breznak) [inline src/nupic/algorithms/SpatialPooler.cpp]: I'd use `while(!activeColumns.empty() && overlaps[activeColumns.back()] < stimulusThreshold_)`
and you could avoid the if-else

File: src/test/unit/algorithms/SpatialPoolerTest.cpp [hunk 3]
```diff
@@ -1108,145 +1108,6 @@ TEST(SpatialPoolerTest, testCalculateOverlapPct) {
...
-  ASSERT_TRUE(sp.isWinner_(score, winners, numWinners));
-  score = 0.1;
-  numWinners = 6;
-  ASSERT_TRUE(!sp.isWinner_(score, winners, numWinners));
-  score = 0.1;
-  numWinners = 7;
-  ASSERT_TRUE(sp.isWinner_(score, winners, numWinners));
-}
-
-TEST(SpatialPoolerTest, testAddToWinners) {
```
Comment (breznak) [inline src/test/unit/algorithms/SpatialPoolerTest.cpp]: ok to remove the code that tests the specific (private) methods. But you should introduce a test that validates the winners functionality

Comment (ctrl-z-9000-times) [reply src/test/unit/algorithms/SpatialPoolerTest.cpp]: Ok, I can fix these things

**Pull request merged**